src/java/org/apache/cassandra/config/DatabaseDescriptor.java

```diff
@@ -4763,4 +4756,159 @@ public static OptionalDouble getSeverityDuringDecommission()
                OptionalDouble.of(conf.severity_during_decommission) :
                OptionalDouble.empty();
     }
+
+    public static void visit(PropertyVisitor visitor)
+    {
+        visit(visitor, RuntimeException::new);
+    }
+
+    public static void visit(PropertyVisitor visitor, Function<Throwable, ? extends RuntimeException> handler)
+    {
+        try
+        {
+            for (Property property : ofNullable(confValueAccessors).orElse(Collections.emptyMap()).values())
+                visitor.visit(property.getName(), property.getType(), property.getAnnotation(Mutable.class) == null);
+        }
+        catch (Throwable t)
+        {
+            throw handler.apply(t);
+        }
+    }
+
+    public static <T> ListenableProperty.Remover addBeforeChangePropertyListener(String name, Class<T> clazz, BiConsumer<T, T> consumer)
+    {
+        return addPropertyListener(name, clazz, p -> p.addBeforeHandler(ListenableProperty.Handler.consume(consumer)));
+    }
+
+    public static <T> ListenableProperty.Remover addAfterChangePropertyListener(String name, Class<T> clazz, BiConsumer<T, T> consumer)
+    {
+        return addPropertyListener(name, clazz, p -> p.addAfterHandler(ListenableProperty.Handler.consume(consumer)));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> ListenableProperty.Remover addPropertyListener(String name,
+                                                                      Class<T> clazz,
+                                                                      Function<ListenableProperty<Config, T>, ListenableProperty.Remover> adder)
+    {
+        Property property = ofNullable(confValueAccessors.get(name))
+                            .map(p -> {
+                                // The case when replacement property is used and match the property's name.
+                                if (p instanceof Replacement.ReplacementProperty)
+                                    return ((Replacement.ReplacementProperty) p).delegate();
+                                else
+                                    return p;
+                            })
+                            .orElseThrow(() -> new PropertyNotFoundException(name));
```

Comment:
can we simplify, trying to use `Optional` actually makes this less clear and cost more

```
if (! confValueAccessors.containsKey(name) new PropertyNotFoundException(name)
Property p = confValueAccessors,get(name);
```

Reply:
Simplified.